File: sdk/go/tipsyabconfig/grpc_target_test.go
package tipsyabconfig

import (
	"strings"
	"testing"
)

// TestParseGRPCTarget_Success is the table-driven coverage of every VALID
// address form in the 方案 Y grammar (design "gRPC 地址 Scheme 解析 + TLS 接入").
//
// The contract under test is the parser's classification, NOT dialing:
//
//	dialTarget         — scheme+query stripped (managed schemes) or verbatim
//	useTLS             — true only for grpcs://
//	authority          — query override (empty = no override)
//	insecureSkipVerify — grpcs://...?insecure=true
//
// The most load-bearing rows are the regression backstops (F1/F9): native
// grpc-go resolver targets (passthrough:///, dns:///, unix:, xds:///) and bare
// host:port MUST be judged "pass-through + plaintext" so all bufconn /
// internal-DNS dialing keeps working unchanged.
func TestParseGRPCTarget_Success(t *testing.T) {
	type want struct {
		dialTarget         string
		useTLS             bool
		authority          string
		insecureSkipVerify bool
	}
	cases := []struct {
		name string
		addr string
		want want
	}{
		// ---- bare host:port → pass-through + plaintext (backward compatible) ----
		{
			name: "bare hostname:port",
			addr: "ab-config-grpc:50051",
			want: want{dialTarget: "ab-config-grpc:50051", useTLS: false},
		},
		{
			// F9 core: 127.0.0.1:443 must NOT be misread (the dotted-quad must
			// not be treated as a scheme); plaintext pass-through.
			name: "bare ipv4:port (F9 regression — not a scheme)",
			addr: "127.0.0.1:443",
			want: want{dialTarget: "127.0.0.1:443", useTLS: false},
		},
		// ---- native grpc-go resolver targets → pass-through + plaintext (F1) ----
		{
			// F1 core: the Go harness uses exactly this address; it contains
			// "://" yet MUST pass through plaintext so every bufconn regression
			// (pull/subscribe/cache/exposure/...) is untouched.
			name: "passthrough:/// (F1 regression — bufconn backstop)",
			addr: "passthrough:///bufnet-config",
			want: want{dialTarget: "passthrough:///bufnet-config", useTLS: false},
		},
		{
			name: "dns:/// native target",
			addr: "dns:///ab-config-grpc.svc.cluster.local:50051",
			want: want{dialTarget: "dns:///ab-config-grpc.svc.cluster.local:50051", useTLS: false},
		},
		{
			name: "unix: native target",
			addr: "unix:/var/run/abconfig.sock",
			want: want{dialTarget: "unix:/var/run/abconfig.sock", useTLS: false},
		},
		{
			name: "xds:/// native target",
			addr: "xds:///ab-config-grpc",
			want: want{dialTarget: "xds:///ab-config-grpc", useTLS: false},
		},
		// ---- grpc:// (explicit plaintext) ----
		{
			name: "grpc:// strips scheme, plaintext",
			addr: "grpc://ab-config-grpc:50051",
			want: want{dialTarget: "ab-config-grpc:50051", useTLS: false},
		},
		// ---- grpcs:// (TLS) ----
		{
			name: "grpcs:// no query → TLS, no authority",
			addr: "grpcs://prod-ab-config-grpc.infra.example.com:443",
			want: want{dialTarget: "prod-ab-config-grpc.infra.example.com:443", useTLS: true},
		},
		{
			// The exact Dev接入串 from the design.
			name: "grpcs:// with authority + insecure=true (Dev串)",
			addr: "grpcs://47.253.175.59:443?authority=dev-ab-config-grpc.infra.fantacy.live&insecure=true",
			want: want{
				dialTarget:         "47.253.175.59:443",
				useTLS:             true,
				authority:          "dev-ab-config-grpc.infra.fantacy.live",
				insecureSkipVerify: true,
			},
		},
		{
			name: "grpcs:// authority only (no insecure)",
			addr: "grpcs://10.0.0.5:443?authority=ab-config-grpc.internal",
			want: want{
				dialTarget: "10.0.0.5:443",
				useTLS:     true,
				authority:  "ab-config-grpc.internal",
			},
		},
		{
			name: "grpcs:// insecure=false is explicit-on-verify",
			addr: "grpcs://host.example.com:443?insecure=false",
			want: want{dialTarget: "host.example.com:443", useTLS: true, insecureSkipVerify: false},
		},
		{
			name: "grpcs:// insecure=1 truthy",
			addr: "grpcs://host.example.com:443?insecure=1",
			want: want{dialTarget: "host.example.com:443", useTLS: true, insecureSkipVerify: true},
		},
		{
			name: "grpcs:// insecure=0 falsy",
			addr: "grpcs://host.example.com:443?insecure=0",
			want: want{dialTarget: "host.example.com:443", useTLS: true, insecureSkipVerify: false},
		},
		// ---- IPv6 ----
		{
			name: "grpcs:// IPv6 literal with port → TLS",
			addr: "grpcs://[::1]:443",
			want: want{dialTarget: "[::1]:443", useTLS: true},
		},
		{
			name: "grpcs:// IPv6 literal with authority",
			addr: "grpcs://[2001:db8::1]:443?authority=ab-config-grpc.internal",
			want: want{
				dialTarget: "[2001:db8::1]:443",
				useTLS:     true,
				authority:  "ab-config-grpc.internal",
			},
		},
	}

	for _, tc := range cases {
		tc := tc
		t.Run(tc.name, func(t *testing.T) {
			got, err := parseGRPCTarget(tc.addr)
			if err != nil {
				t.Fatalf("parseGRPCTarget(%q) unexpected error: %v", tc.addr, err)
			}
			if got.dialTarget != tc.want.dialTarget {
				t.Errorf("dialTarget = %q, want %q", got.dialTarget, tc.want.dialTarget)
			}
			if got.useTLS != tc.want.useTLS {
				t.Errorf("useTLS = %v, want %v", got.useTLS, tc.want.useTLS)
			}
			if got.authority != tc.want.authority {
				t.Errorf("authority = %q, want %q", got.authority, tc.want.authority)
			}
			if got.insecureSkipVerify != tc.want.insecureSkipVerify {
				t.Errorf("insecureSkipVerify = %v, want %v", got.insecureSkipVerify, tc.want.insecureSkipVerify)
			}
		})
	}
}

// TestParseGRPCTarget_Errors covers every ERROR path. All are parameter errors
// (prefix "tipsyabconfig:") surfaced at parse time. We assert an error is
// returned and that its message contains a discriminating keyword so a future
// reword that drops the contract intent is caught.
func TestParseGRPCTarget_Errors(t *testing.T) {
	cases := []struct {
		name        string
		addr        string
		wantSubstrs []string // all must appear (case-sensitive) somewhere in err
	}{
		{
			// Q1: query on a plaintext grpc:// target is meaningless → error.
			name:        "grpc:// with authority query rejected",
			addr:        "grpc://ab-config-grpc:50051?authority=x",
			wantSubstrs: []string{"query parameters are only valid"},
		},
		{
			name:        "grpc:// with insecure query rejected",
			addr:        "grpc://ab-config-grpc:50051?insecure=true",
			wantSubstrs: []string{"query parameters are only valid"},
		},
		{
			// Q2: grpcs:// without an explicit port is an error (no implicit :443).
			name:        "grpcs:// missing port",
			addr:        "grpcs://ab-config-grpc.internal",
			wantSubstrs: []string{"explicit port"},
		},
		{
			name:        "grpcs:// IPv6 missing port",
			addr:        "grpcs://[::1]",
			wantSubstrs: []string{"explicit port"},
		},
		{
			// S2: a present-but-non-numeric port is "invalid port", distinct from
			// the missing-port "explicit port" error above and surfaced at parse
			// time (not deferred to the dialer).
			name:        "grpcs:// non-numeric port",
			addr:        "grpcs://host:abc",
			wantSubstrs: []string{"invalid port"},
		},
		{
			name:        "grpcs:// IPv6 non-numeric port",
			addr:        "grpcs://[::1]:abc",
			wantSubstrs: []string{"invalid port"},
		},
		{
			name:        "grpcs:// out-of-range port",
			addr:        "grpcs://host:99999",
			wantSubstrs: []string{"invalid port"},
		},
		{
			name:        "grpcs:// unknown query key rejected",
			addr:        "grpcs://host.example.com:443?foo=bar",
			wantSubstrs: []string{"unknown query parameter"},
		},
		{
			name:        "grpcs:// invalid insecure value rejected",
			addr:        "grpcs://host.example.com:443?insecure=maybe",
			wantSubstrs: []string{"invalid insecure value"},
		},
		{
			// design rule 5: http:// in gRPC mode is a parameter error.
			name:        "http:// in gRPC mode rejected",
			addr:        "http://lb.internal:8080",
			wantSubstrs: []string{"HTTP base URL"},
		},
		{
			name:        "https:// in gRPC mode rejected",
			addr:        "https://lb.internal:8443",
			wantSubstrs: []string{"HTTP base URL"},
		},
	}

	for _, tc := range cases {
		tc := tc
		t.Run(tc.name, func(t *testing.T) {
			got, err := parseGRPCTarget(tc.addr)
			if err == nil {
				t.Fatalf("parseGRPCTarget(%q) = %+v, want error", tc.addr, got)
			}
			msg := err.Error()
			if !strings.HasPrefix(msg, "tipsyabconfig:") {
				t.Errorf("error should be a tipsyabconfig: parameter error, got %q", msg)
			}
			for _, sub := range tc.wantSubstrs {
				if !strings.Contains(msg, sub) {
					t.Errorf("error %q does not contain expected keyword %q", msg, sub)
				}
			}
		})
	}
}

// TestParseGRPCTarget_PassthroughIsPlaintext is an explicit, standalone
// assertion of the single most important regression backstop (design Risks /
// F1): the bufconn harness address must parse to "pass-through + plaintext"
// with no authority and no skip-verify. Kept separate from the table so a
// breakage names this exact contract.
func TestParseGRPCTarget_PassthroughIsPlaintext(t *testing.T) {
	for _, addr := range []string{
		"passthrough:///bufnet-config",
		"passthrough:///bufnet-abtest",
	} {
		got, err := parseGRPCTarget(addr)
		if err != nil {
			t.Fatalf("parseGRPCTarget(%q): unexpected error %v", addr, err)
		}
		if got.dialTarget != addr {
			t.Errorf("dialTarget = %q, want verbatim %q", got.dialTarget, addr)
		}
		if got.useTLS {
			t.Errorf("useTLS = true for %q; native resolver targets must stay plaintext", addr)
		}
		if got.authority != "" {
			t.Errorf("authority = %q for %q; must be empty (no override)", got.authority, addr)
		}
		if got.insecureSkipVerify {
			t.Errorf("insecureSkipVerify = true for %q; must be false", addr)
		}
	}
}

// TestParseGRPCTarget_BareLoopbackIsPlaintext pins the other F9 backstop: a
// bare 127.0.0.1:443 (the form Python's conftest uses; symmetric here for the
// Go contract) is plaintext pass-through, never misclassified as TLS.
func TestParseGRPCTarget_BareLoopbackIsPlaintext(t *testing.T) {
	got, err := parseGRPCTarget("127.0.0.1:443")
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if got.useTLS || got.dialTarget != "127.0.0.1:443" {
		t.Fatalf("bare 127.0.0.1:443 misclassified: %+v", got)
	}
}

File: sdk/java/tipsy-abconfig/src/test/java/io/github/lightspeedintelligence/abconfig/GrpcTargetTest.java
package io.github.lightspeedintelligence.abconfig;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.junit.jupiter.api.Assertions.assertTrue;

import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Nested;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.Arguments;
import org.junit.jupiter.params.provider.CsvSource;
import org.junit.jupiter.params.provider.MethodSource;
import org.junit.jupiter.params.provider.ValueSource;

import java.util.stream.Stream;

/**
 * ST5 — {@link GrpcTarget#parseGrpcTarget(String)} 方案 Y 地址解析矩阵。
 *
 * <p>语义基准是 Go {@code grpc_target.go} / {@code grpc_target_test.go}：被测的契约
 * 是 <em>分类</em>（dialTarget / useTls / authority / insecureSkipVerify），不是 dial。
 *
 * <p>最关键的回归 backstop（F1/F9）：原生 grpc resolver 目标
 * （{@code passthrough:///}、{@code dns:///}、{@code unix:}、{@code xds:///}）与 bare
 * {@code host:port} 必须被判为「透传 + 明文」，使既有 bufconn / 内部 DNS dialing 不受影响。
 * 所有错误为参数错误（{@link ConfigValidationException}，消息前缀 {@code tipsyabconfig:}），
 * 在建 channel 前抛出。
 */
class GrpcTargetTest {

    // ------------------------------------------------------------------
    // 成功分类矩阵（每行：addr -> dialTarget/useTls/authority/insecureSkipVerify）
    // ------------------------------------------------------------------

    static Stream<Arguments> successCases() {
        return Stream.of(
                // ---- bare host:port -> 透传 + 明文（向后兼容） ----
                Arguments.of("bare hostname:port",
                        "ab-config-grpc:50051", "ab-config-grpc:50051", false, "", false),
                // F9 核心：127.0.0.1:443 不能被误读为 scheme；明文透传。
                Arguments.of("bare ipv4:port (F9 — 非 scheme)",
                        "127.0.0.1:443", "127.0.0.1:443", false, "", false),

                // ---- 原生 grpc resolver 目标 -> 透传 + 明文（F1） ----
                // F1 核心：含 "://" 但必须明文透传（bufnet bufconn backstop）。
                Arguments.of("passthrough:/// (F1 — bufconn backstop)",
                        "passthrough:///bufnet", "passthrough:///bufnet", false, "", false),
                Arguments.of("dns:/// 原生目标",
                        "dns:///svc.ns:443", "dns:///svc.ns:443", false, "", false),
                Arguments.of("unix: 原生目标",
                        "unix:/tmp/x", "unix:/tmp/x", false, "", false),
                Arguments.of("xds:/// 原生目标",
                        "xds:///foo", "xds:///foo", false, "", false),

                // ---- grpc:// 显式明文 ----
                Arguments.of("grpc:// 剥 scheme, 明文",
                        "grpc://host:50051", "host:50051", false, "", false),

                // ---- grpcs:// (TLS) ----
                Arguments.of("grpcs:// 无 query -> TLS, 无 authority",
                        "grpcs://host:443", "host:443", true, "", false),
                // 设计 Dev 接入串：authority + insecure=true。
                Arguments.of("grpcs:// authority + insecure=true (Dev串)",
                        "grpcs://host:443?authority=dom&insecure=true",
                        "host:443", true, "dom", true),
                Arguments.of("grpcs:// authority only (无 insecure)",
                        "grpcs://10.0.0.5:443?authority=ab-config-grpc.internal",
                        "10.0.0.5:443", true, "ab-config-grpc.internal", false),
                // insecure 取值 true/1/false/0 各对。
                Arguments.of("grpcs:// insecure=true",
                        "grpcs://host:443?insecure=true", "host:443", true, "", true),
                Arguments.of("grpcs:// insecure=1 truthy",
                        "grpcs://host:443?insecure=1", "host:443", true, "", true),
                Arguments.of("grpcs:// insecure=false explicit-verify",
                        "grpcs://host:443?insecure=false", "host:443", true, "", false),
                Arguments.of("grpcs:// insecure=0 falsy",
                        "grpcs://host:443?insecure=0", "host:443", true, "", false),

                // ---- IPv6 字面量 ----
                Arguments.of("grpcs:// IPv6 字面量 + port -> TLS, host 解析正确",
                        "grpcs://[::1]:443", "[::1]:443", true, "", false),
                Arguments.of("grpcs:// IPv6 字面量 + authority",
                        "grpcs://[2001:db8::1]:443?authority=ab-config-grpc.internal",
                        "[2001:db8::1]:443", true, "ab-config-grpc.internal", false));
    }

    @ParameterizedTest(name = "[{index}] {0}")
    @MethodSource("successCases")
    @DisplayName("parseGrpcTarget: 合法地址 -> 正确分类")
    void parseSuccess(String name, String addr, String wantDial, boolean wantTls,
            String wantAuthority, boolean wantInsecure) {
        GrpcTarget got = GrpcTarget.parseGrpcTarget(addr);
        assertEquals(wantDial, got.dialTarget(), "dialTarget for " + addr);
        assertEquals(wantTls, got.useTls(), "useTls for " + addr);
        assertEquals(wantAuthority, got.authority(), "authority for " + addr);
        assertEquals(wantInsecure, got.insecureSkipVerify(), "insecureSkipVerify for " + addr);
    }

    // ------------------------------------------------------------------
    // 错误矩阵：所有路径抛 ConfigValidationException，前缀 tipsyabconfig:，
    // 且消息含可区分关键字（捕捉「重写时丢掉契约意图」）。
    // ------------------------------------------------------------------

    static Stream<Arguments> errorCases() {
        return Stream.of(
                // grpc:// 带 query（authority / insecure）-> 参数错误。
                Arguments.of("grpc:// 带 authority query 拒绝",
                        "grpc://host:50051?authority=x", "query parameters are only valid"),
                Arguments.of("grpc:// 带 insecure query 拒绝",
                        "grpc://host:50051?insecure=true", "query parameters are only valid"),

                // grpcs:// 缺端口 -> "explicit port"（与 invalid port 区分）。
                Arguments.of("grpcs:// 缺端口",
                        "grpcs://host", "explicit port"),
                Arguments.of("grpcs:// IPv6 缺端口",
                        "grpcs://[::1]", "explicit port"),

                // grpcs:// 非数字 / 越界端口 -> "invalid port"。
                Arguments.of("grpcs:// 非数字端口",
                        "grpcs://host:abc", "invalid port"),
                Arguments.of("grpcs:// IPv6 非数字端口",
                        "grpcs://[::1]:abc", "invalid port"),
                Arguments.of("grpcs:// 越界端口 99999",
                        "grpcs://host:99999", "invalid port"),

                // grpcs:// 缺 host（:443）-> 缺 host 报错。
                Arguments.of("grpcs:// 缺 host (:443)",
                        "grpcs://:443", "missing a host"),

                // grpcs:// 未知 query key / 非法 insecure 值。
                Arguments.of("grpcs:// 未知 query key 拒绝",
                        "grpcs://host:443?foo=bar", "unknown query parameter"),
                Arguments.of("grpcs:// 非法 insecure 值拒绝",
                        "grpcs://host:443?insecure=maybe", "invalid insecure value"),

                // http:// / https:// 在 gRPC 模式 -> 提示用 Transport=http。
                Arguments.of("http:// 在 gRPC 模式拒绝",
                        "http://lb.internal:8080", "HTTP base URL"),
                Arguments.of("https:// 在 gRPC 模式拒绝",
                        "https://lb.internal:8443", "HTTP base URL"));
    }

    @ParameterizedTest(name = "[{index}] {0}")
    @MethodSource("errorCases")
    @DisplayName("parseGrpcTarget: 非法地址 -> ConfigValidationException(tipsyabconfig: + 关键字)")
    void parseError(String name, String addr, String wantKeyword) {
        ConfigValidationException ex = assertThrows(ConfigValidationException.class,
                () -> GrpcTarget.parseGrpcTarget(addr), "expected error for " + addr);
        String msg = ex.getMessage();
        assertTrue(msg.startsWith("tipsyabconfig:"),
                "message must carry tipsyabconfig: prefix, got: " + msg);
        assertTrue(msg.contains(wantKeyword),
                "message must contain keyword \"" + wantKeyword + "\", got: " + msg);
    }

    // ------------------------------------------------------------------
    // 显式 backstop：缺端口 vs 非数字端口必须是两个不同错误（设计明确区分）。
    // ------------------------------------------------------------------

    @Test
    @DisplayName("缺端口与非数字端口报不同错误关键字（explicit port vs invalid port）")
    void missingPortVsInvalidPortAreDistinct() {
        ConfigValidationException missing = assertThrows(ConfigValidationException.class,
                () -> GrpcTarget.parseGrpcTarget("grpcs://host"));
        ConfigValidationException invalid = assertThrows(ConfigValidationException.class,
                () -> GrpcTarget.parseGrpcTarget("grpcs://host:abc"));

        assertTrue(missing.getMessage().contains("explicit port"),
                "缺端口应报 explicit port: " + missing.getMessage());
        assertFalse(missing.getMessage().contains("invalid port"),
                "缺端口不应误报 invalid port: " + missing.getMessage());

        assertTrue(invalid.getMessage().contains("invalid port"),
                "非数字端口应报 invalid port: " + invalid.getMessage());
        assertFalse(invalid.getMessage().contains("explicit port"),
                "非数字端口不应误报 explicit port: " + invalid.getMessage());
    }

    // ------------------------------------------------------------------
    // ConfigValidationException 是 TipsyConfigException 的子类（异常体系）。
    // ------------------------------------------------------------------

    @Test
    @DisplayName("解析错误为 ConfigValidationException 且是 TipsyConfigException 子类")
    void errorIsConfigValidationSubclassOfTipsyConfig() {
        ConfigValidationException ex = assertThrows(ConfigValidationException.class,
                () -> GrpcTarget.parseGrpcTarget("http://x"));
        assertTrue(ex instanceof TipsyConfigException,
                "ConfigValidationException 必须是 TipsyConfigException 子类");
    }

    // ------------------------------------------------------------------
    // 标准化的 F1 / F9 backstop（独立命名，便于断点定位回归）。
    // ------------------------------------------------------------------

    @Nested
    @DisplayName("回归 backstop（F1/F9）")
    class RegressionBackstops {

        @ParameterizedTest(name = "passthrough 透传明文: {0}")
        @ValueSource(strings = {"passthrough:///bufnet-config", "passthrough:///bufnet-abtest"})
        @DisplayName("passthrough:/// 必须透传明文（F1 — bufconn backstop）")
        void passthroughIsPlaintext(String addr) {
            GrpcTarget got = GrpcTarget.parseGrpcTarget(addr);
            assertEquals(addr, got.dialTarget(), "dialTarget 必须逐字透传");
            assertFalse(got.useTls(), "原生 resolver 目标必须保持明文");
            assertEquals("", got.authority(), "authority 必须为空（不覆盖）");
            assertFalse(got.insecureSkipVerify(), "insecureSkipVerify 必须为 false");
        }

        @Test
        @DisplayName("bare 127.0.0.1:443 明文透传（F9 — 不误判为 TLS）")
        void bareLoopbackIsPlaintext() {
            GrpcTarget got = GrpcTarget.parseGrpcTarget("127.0.0.1:443");
            assertFalse(got.useTls(), "bare 127.0.0.1:443 不应被误判为 TLS");
            assertEquals("127.0.0.1:443", got.dialTarget());
        }
    }

    // ------------------------------------------------------------------
    // insecure 取值的 CSV 矩阵（与 MethodSource 重叠，但用 CsvSource 保证
    // 真值表完整、断点定位单值）。
    // ------------------------------------------------------------------

    @ParameterizedTest(name = "insecure={0} -> {1}")
    @CsvSource({
            "true,  true",
            "1,     true",
            "false, false",
            "0,     false"
    })
    @DisplayName("grpcs:// insecure 真值表 true/1/false/0")
    void insecureTruthTable(String raw, boolean expected) {
        GrpcTarget got = GrpcTarget.parseGrpcTarget("grpcs://host:443?insecure=" + raw);
        assertTrue(got.useTls());
        assertEquals(expected, got.insecureSkipVerify(),
                "insecure=" + raw + " should map to " + expected);
    }
}
